homeassistant/components/yolink/__init__.py

```diff
@@ -49,12 +58,83 @@
 class YoLinkCoordinator(DataUpdateCoordinator):
     """YoLink DataUpdateCoordinator."""
 
+    yl_devices: list[YoLinkDevice] = []
+
     def __init__(
-        self,
-        hass: HomeAssistant,
+        self, hass: HomeAssistant, yl_client: YoLinkClient, yl_mqtt_client: MqttClient
     ) -> None:
-        """Init YoLink DataUpdateCoordinator."""
-        super().__init__(hass, _LOGGER, name=DOMAIN)
+        """Init YoLink DataUpdateCoordinator.
+
+        fetch state every 30 minutes base on yolink device heartbeat interval
+        """
+        super().__init__(
+            hass, _LOGGER, name=DOMAIN, update_interval=timedelta(minutes=30)
+        )
+        self._client = yl_client
+        self._mqtt_client = yl_mqtt_client
+        self.data = {}
+
+    async def init_coordinator(self):
+        """Init coordinator."""
+        try:
+            async with async_timeout.timeout(10):
+                home_info = await self._client.get_general_info()
+
+                def on_message_callback(message: tuple[str, BRDP]):
```

Comment:
It doesn't seem we need to nest this function here. We can define it as a helper method on the coordinator.